universe/ValueRef.h

```diff
@@ -9,10 +9,37 @@
 
 namespace ValueRef {
 
-/** The base class for all ValueRef classes.  This class provides the public
-  * interface for a ValueRef expression tree. */
+//! The common base class for all ValueRef classes. This class provides
+//! some the return-type-independent interface.
+struct FO_COMMON_API ValueRefBase {
```

Comment:
This class is named horrible and I doubt that it should exist in the first place. Why do you do that? Why would you need that?

Reply:
So different `ValueRef<T>` can have a common base interface and be stored in a simple container of pointers. This allows invariance tests on a range of valuerefs of different types (int, string, double) to be simplified with a single range `all_of` call.

It could also be used as a base class / interface for `Condition` so they could also be stored in the same container with valuerefs.

It would probably also allow some other code that passes around a `ValueRef` but doesn't need to know what type is returned to do so with a less bulky #include 

It also [avoids redundant instantiation](https://books.google.de/books?id=m2cnCgAAQBAJ&pg=PA19) of all of the functions that were moved from the templated class into the non-templated base class, which [don't depend on the template](https://blog.feabhas.com/2014/06/template-inheritance/) type (eg. simple getters of non-template-dependent member bool flags).

Reply:
@adrianbroher any further comments about this?

Reply:
> So different ValueRef<T> can have a common base interface and be stored in a simple container of pointers-to-base. This allows invariance tests on a range of valuerefs of different types (int, string, double) to be simplified with a single range all_of call.

> It would probably also allow some other code that passes around a ValueRef but doesn't need to know what type is returned to do so with a less bulky #include

And then you have to distinguish them again with expensive casts when you want to evaluate the value of the expression.  I don't see any gain here.  `ValueRef`s are always bound to the domain with a certain type (e.g. the ships hull structure is always expressed as floating point number, a systems star type is expressed as `StarType` enum, etc).  There is no common ground between those in their nature that could be abstracted out.

> It could also be used as a base class / interface for Condition so they could also be stored in the same container with valuerefs, further simplifying the invariance determination code.

If you want to avoid duplicated code because some common semantic exists, which is not backed by the domain a mixin would be the tool of choice.  This is different to a base class as it encapsulates a concept away without using a base class for, by domain, unrelated classes.

> The name can be changed, but I think ValueRefBase sums it up pretty well, though not if it is also used as a base for Condition.

`focs::Expression` but as mentioned before the idea of a typeless expression is pointless.

Reply:
> > It would probably also allow some other code that passes around a ValueRef but doesn't need to know what type is returned to do so with a less bulky #include
> 
> And then you have to distinguish them again with expensive casts when you want to evaluate the value of the expression.

That would be a case where you do need to know what the type is. If you need to evaluate a `ValueRef`, you probably wouldn't pass it as a base non-evaluable type pointer, and that's not what I'm referring to.

> I don't see any gain here. `ValueRef`s are always bound to the domain with a certain type (e.g. the ships hull structure is always expressed as floating point number, a systems star type is expressed as `StarType` enum, etc). There is no common ground between those in their nature that could be abstracted out.

This pull request contains examples in which several `ValueRef`s of different types are passed and treated the same way and where it's convenient to store them in a single container. The `ValueRefBase` class shows the commonality between `ValueRef`s of different type specialization (the functions and storage that don't depend on the template type) which is used by that code. Specifically: when determining if a `Condition` or `ValueRef` that has subconditions or subvaluerefs is Source/Target/LocalCandidate/RootCandidate invariant, the template type doesn't matter and the `ValueRef`s aren't evaluated.

It doesn't matter if the sub-expressions are `Condition` or `ValueRef<T>` or what `T` is; these all share the same invariance properties. That the `Condition` or `ValueRef<T>` can be evaluated on object sets or to get a number, string, or enum doesn't matter in that code, as that code doesn't evaluate them.

> > It could also be used as a base class / interface for Condition so they could also be stored in the same container with valuerefs, further simplifying the invariance determination code.
> 
> If you want to avoid duplicated code because some common semantic exists, which is not backed by the domain a mixin would be the tool of choice. This is different to a base class as it encapsulates a concept away without using a base class for, by domain, unrelated classes.

So you propose having an `InvarianceInfo` struct that `Condition` and `ValueRef<T>` have member variables of, and when doing the invariance tests, use a list of pointers to `InvarianceInfo`, which is initialized by `value_ref->GetInvarianceInfo()` and `condition->GetInvarianceInfo()` calls? I suppose that could work, but it seems awkward compared to storing the base class pointers. It also doesn't avoid instantiating the common invariance getters that don't depend on the template type, which was another advantage of the common base class pattern.

Reply:
So whatever you want then.  I will push an improvement PR then when I have time.